// @flixora/zeptpay-payment-core/dist/validation/AirXPayValid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateSimulateQR = exports.validateCreateAutoPay = exports.validateGenerateQR = exports.validateCreatePayment = exports.validatePayer = exports.ValidationError = void 0;
class ValidationError extends Error {
    constructor(field, message) {
        super(message);
        this.field = field;
        this.name = "ValidationError";
    }
}
exports.ValidationError = ValidationError;
const assertPresent = (value, fieldName) => {
    if (value === undefined || value === null || value === "") {
        throw new ValidationError(fieldName, `${fieldName} is required`);
    }
};
const assertPositiveNumber = (value, fieldName) => {
    if (typeof value !== "number" || value <= 0) {
        throw new ValidationError(fieldName, `${fieldName} must be a positive number`);
    }
};
const validatePayer = (payer) => {
    assertPresent(payer, "payer");
    assertPresent(payer === null || payer === void 0 ? void 0 : payer.userId, "payer.userId");
    assertPresent(payer === null || payer === void 0 ? void 0 : payer.name, "payer.name");
    assertPresent(payer === null || payer === void 0 ? void 0 : payer.email, "payer.email");
};
exports.validatePayer = validatePayer;
const validateCreatePayment = (data) => {
    assertPresent(data.vendorCodeUID, "vendorCodeUID");
    assertPresent(data.appName, "appName");
    assertPositiveNumber(data.amount, "amount");
    (0, exports.validatePayer)(data.payer);
};
exports.validateCreatePayment = validateCreatePayment;
const validateGenerateQR = (data) => {
    assertPresent(data.vendorCodeUID, "vendorCodeUID");
    assertPresent(data.appName, "appName");
    assertPositiveNumber(data.amount, "amount");
    (0, exports.validatePayer)(data.payer);
};
exports.validateGenerateQR = validateGenerateQR;
const validateCreateAutoPay = (data) => {
    assertPresent(data.vendorCodeUID, "vendorCodeUID");
    assertPresent(data.appName, "appName");
    assertPositiveNumber(data.amount, "amount");
    (0, exports.validatePayer)(data.payer);
    const validFrequencies = ["daily", "weekly", "monthly"];
    if (data.frequency && !validFrequencies.includes(data.frequency)) {
        throw new ValidationError("frequency", `frequency must be one of: ${validFrequencies.join(", ")}`);
    }
};
exports.validateCreateAutoPay = validateCreateAutoPay;
const validateSimulateQR = (data) => {
    assertPresent(data.qrCodeId, "qrCodeId");
    assertPresent(data.result, "result");
    const validResults = ["success", "failed", "pending"];
    if (!validResults.includes(data.result)) {
        throw new ValidationError("result", `result must be one of: ${validResults.join(", ")}`);
    }
};
exports.validateSimulateQR = validateSimulateQR;
