// @flixora/zeptpay-payment-core/dist/modules/payments.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentsModule = void 0;
const ZeptPayValid_1 = require("../validation/ZeptPayValid");
class PaymentsModule {
    constructor(client) {
        this.client = client;
    }
    createPayment(data) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, ZeptPayValid_1.validateCreatePayment)(data);
            return this.client.post("/api/test-live/payments", data);
        });
    }
    getTransaction(transactionId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!transactionId)
                throw new Error("transactionId is required");
            return this.client.get(`/api/test-live/payments/transaction/${transactionId}`);
        });
    }
    getUserTransactions(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { userId, role, limit, skip } = params;
            if (!userId)
                throw new Error("userId is required");
            return this.client.get(`/api/test-live/payments/user/${userId}/transactions`, { role, limit, skip });
        });
    }
    confirmPayment(data) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!data.txnId)
                throw new Error("txnId is required");
            return this.client.post("/api/test-live/payments/confirm", data);
        });
    }
}
exports.PaymentsModule = PaymentsModule;
