// @flixora/zeptpay-payment-core/dist/modules/zeptpayWebhook.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeptPayWebhook = void 0;
// sdk/src/modules/zeptpayWebhook.ts
const crypto_1 = __importDefault(require("crypto"));
class ZeptPayWebhook {
    static init(client) {
        this.client = client;
    }
    static verifyEventZPTFXO(payload, signature, timestamp, webhookKey) {
        if (!webhookKey || !webhookKey.startsWith("whsec-")) {
            throw new Error("Invalid webhook key format. Must start with 'whsec-'");
        }
        if (!signature || !timestamp) {
            throw new Error("Missing webhook headers: signature and timestamp required");
        }
        const timestampNum = parseInt(timestamp, 10);
        const now = Math.floor(Date.now() / 1000);
        if (isNaN(timestampNum) || Math.abs(now - timestampNum) > 300) {
            throw new Error("Webhook timestamp too old or invalid");
        }
        const payloadString = JSON.stringify(payload);
        const signaturePayload = `${timestampNum}.${payloadString}`;
        const expectedSignature = crypto_1.default
            .createHmac("sha256", webhookKey)
            .update(signaturePayload)
            .digest("hex");
        if (expectedSignature !== signature) {
            throw new Error("Invalid webhook signature");
        }
        return {
            id: payload.id,
            type: payload.type,
            data: payload.data,
            created: payload.created,
            raw: payload,
        };
    }
    static getStatus(webhookKey) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.client) {
                throw new Error("ZeptPayWebhook not initialized");
            }
            return this.client.get(`/api/webhook/status/${webhookKey}`);
        });
    }
}
exports.ZeptPayWebhook = ZeptPayWebhook;
