// @flixora/zeptpay-payment-core/dist/clients/AirXPayClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeptPayClient = void 0;
const axios_1 = __importDefault(require("axios"));
const ZeptPayOption_1 = require("../options/ZeptPayOption");
const errorHandler_1 = require("../error/errorHandler");
const PerformanceTracker_1 = require("../performance/PerformanceTracker");
const EventEmitter_1 = require("../event/EventEmitter");
const resolveConfig = (config) => {
    var _a, _b;
    if (!config.clientKey)
        throw new Error("clientKey is required");
    if (!config.secretKey)
        throw new Error("secretKey is required");
    return {
        clientKey: config.clientKey,
        secretKey: config.secretKey,
        timeout: (_a = config.timeout) !== null && _a !== void 0 ? _a : ZeptPayOption_1.DEFAULT_TIMEOUT,
        testMode: (_b = config.testMode) !== null && _b !== void 0 ? _b : true,
    };
};
class ZeptPayClient {
    constructor(config) {
        this.config = resolveConfig(config);
        this.events = new EventEmitter_1.EventEmitter();
        this.performance = new PerformanceTracker_1.PerformanceTracker();
        this.axiosInstance = axios_1.default.create({
            baseURL: ZeptPayOption_1.ZEPTPAY_BASE_URL, // 🔒 Hidden from developer
            timeout: this.config.timeout,
            headers: {
                "Content-Type": "application/json",
                "x-client-key": this.config.clientKey,
                "x-secret-key": this.config.secretKey,
                "x-sdk-name": ZeptPayOption_1.SDK_NAME,
                "x-sdk-version": ZeptPayOption_1.SDK_VERSION,
            },
        });
        this._setupInterceptors();
    }
    _setupInterceptors() {
        this.axiosInstance.interceptors.request.use((req) => {
            this.events.emit("request", { url: req.url, method: req.method });
            return req;
        });
        this.axiosInstance.interceptors.response.use((res) => {
            this.events.emit("response", {
                status: res.status,
                url: res.config.url,
            });
            return res;
        }, (err) => {
            this.events.emit("error", { error: err.message, code: err.code });
            return Promise.reject(err);
        });
    }
    post(url, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const start = Date.now();
            try {
                const res = yield this.axiosInstance.post(url, data);
                this.performance.track(url, Date.now() - start);
                return res.data;
            }
            catch (err) {
                return (0, errorHandler_1.handleAxiosError)(err);
            }
        });
    }
    get(url, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const start = Date.now();
            try {
                const res = yield this.axiosInstance.get(url, {
                    params,
                });
                this.performance.track(url, Date.now() - start);
                return res.data;
            }
            catch (err) {
                return (0, errorHandler_1.handleAxiosError)(err);
            }
        });
    }
}
exports.ZeptPayClient = ZeptPayClient;
