// @flixora/zeptpay-payment-core/dist/clients/ZeptPayClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeptPayClient = void 0;
const axios_1 = __importDefault(require("axios"));
const ZeptPayOption_1 = require("../options/ZeptPayOption");
const errorHandler_1 = require("../error/errorHandler");
const PerformanceTracker_1 = require("../performance/PerformanceTracker");
const EventEmitter_1 = require("../event/EventEmitter");
const resolveConfig = (config) => {
    var _a, _b;
    // 🔥 Only validate for API calls, not for webhook-only mode
    // Webhook-only mode: no clientKey/secretKey needed
    return {
        clientKey: config.clientKey,
        secretKey: config.secretKey,
        webhookKey: config.webhookKey,
        timeout: (_a = config.timeout) !== null && _a !== void 0 ? _a : ZeptPayOption_1.DEFAULT_TIMEOUT,
        testMode: (_b = config.testMode) !== null && _b !== void 0 ? _b : true,
    };
};
class ZeptPayClient {
    constructor(config) {
        this.config = resolveConfig(config);
        this.events = new EventEmitter_1.EventEmitter();
        this.performance = new PerformanceTracker_1.PerformanceTracker();
        // 🔥 Check if this is webhook-only mode
        this.isWebhookOnly =
            !this.config.clientKey &&
                !this.config.secretKey &&
                !!this.config.webhookKey;
        // 🔥 Build headers dynamically
        const headers = {
            "Content-Type": "application/json",
            "x-sdk-name": ZeptPayOption_1.SDK_NAME,
            "x-sdk-version": ZeptPayOption_1.SDK_VERSION,
        };
        // 🔥 Only add client keys if provided (for API calls)
        if (this.config.clientKey) {
            headers["x-client-key"] = this.config.clientKey;
        }
        if (this.config.secretKey) {
            headers["x-secret-key"] = this.config.secretKey;
        }
        if (this.config.webhookKey) {
            headers["x-zeptpay-webhook-key"] = this.config.webhookKey;
        }
        this.axiosInstance = axios_1.default.create({
            baseURL: ZeptPayOption_1.ZEPTPAY_BASE_URL,
            timeout: this.config.timeout,
            headers,
        });
        this._setupInterceptors();
        if (this.isWebhookOnly) {
            console.log("🔗 ZeptPayClient initialized in webhook-only mode");
        }
    }
    _setupInterceptors() {
        this.axiosInstance.interceptors.request.use((req) => {
            this.events.emit("request", { url: req.url, method: req.method });
            return req;
        });
        this.axiosInstance.interceptors.response.use((res) => {
            this.events.emit("response", {
                status: res.status,
                url: res.config.url,
            });
            return res;
        }, (err) => {
            this.events.emit("error", { error: err.message, code: err.code });
            return Promise.reject(err);
        });
    }
    post(url, data) {
        return __awaiter(this, void 0, void 0, function* () {
            // 🔥 Prevent API calls in webhook-only mode
            if (this.isWebhookOnly && !url.includes("/webhook/")) {
                throw new Error("API calls require clientKey and secretKey. Use webhook-only mode for webhook verification only.");
            }
            const start = Date.now();
            try {
                const res = yield this.axiosInstance.post(url, data);
                this.performance.track(url, Date.now() - start);
                return res.data;
            }
            catch (err) {
                throw (0, errorHandler_1.handleAxiosError)(err);
            }
        });
    }
    get(url, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const start = Date.now();
            try {
                const res = yield this.axiosInstance.get(url, {
                    params,
                });
                this.performance.track(url, Date.now() - start);
                return res.data;
            }
            catch (err) {
                throw (0, errorHandler_1.handleAxiosError)(err);
            }
        });
    }
    isWebhookOnlyMode() {
        return this.isWebhookOnly;
    }
}
exports.ZeptPayClient = ZeptPayClient;
