// @flixora/zeptpay-payment-core/dist/performance/PerformanceTracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerformanceTracker = void 0;
class PerformanceTracker {
    constructor() {
        this.entries = [];
        this.maxEntries = 100;
    }
    track(url, duration) {
        this.entries.push({ url, duration, timestamp: new Date().toISOString() });
        if (this.entries.length > this.maxEntries)
            this.entries.shift();
    }
    getAll() {
        return [...this.entries];
    }
    getAverageDuration(url) {
        const filtered = url
            ? this.entries.filter((e) => e.url === url)
            : this.entries;
        if (!filtered.length)
            return 0;
        return filtered.reduce((sum, e) => sum + e.duration, 0) / filtered.length;
    }
    clear() {
        this.entries = [];
    }
    getSummary() {
        return {
            totalRequests: this.entries.length,
            averageDuration: this.getAverageDuration(),
            slowestRequest: this.entries.reduce((max, e) => { var _a; return (e.duration > ((_a = max === null || max === void 0 ? void 0 : max.duration) !== null && _a !== void 0 ? _a : 0) ? e : max); }, this.entries[0]),
        };
    }
}
exports.PerformanceTracker = PerformanceTracker;
