// @flixora/zeptpay-payment-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeptPay = void 0;
const ZeptPayClient_1 = require("./clients/ZeptPayClient");
const payments_1 = require("./modules/payments");
const qr_1 = require("./modules/qr");
const autopay_1 = require("./modules/autopay");
const zeptpayWebhook_1 = require("./modules/zeptpayWebhook");
const PluginManager_1 = require("./plugin/PluginManager");
const DevLogger_1 = require("./dev/DevLogger");
const RateLimiter_1 = require("./rule/RateLimiter");
class ZeptPay {
    constructor(config) {
        this.client = new ZeptPayClient_1.ZeptPayClient(config);
        this.logger = new DevLogger_1.DevLogger(process.env.NODE_ENV === "development");
        this.plugins = new PluginManager_1.PluginManager();
        this.rateLimiter = new RateLimiter_1.RateLimiter();
        // Initialize webhook with client
        zeptpayWebhook_1.ZeptPayWebhook.init(this.client);
        // Check if we have API keys
        this.hasApiKeys = !!(config.clientKey && config.secretKey);
        const hasWebhookKey = !!config.webhookKey;
        if (!this.hasApiKeys && !hasWebhookKey) {
            throw new Error("Either clientKey+secretKey (for API) or webhookKey (for webhook) is required");
        }
        // 🔥 flixora namespace - API modules + webhook
        this.flixora = {
            payments: new payments_1.PaymentsModule(this.client),
            qr: new qr_1.QRModule(this.client),
            autopay: new autopay_1.AutoPayModule(this.client),
            webhook: {
                verifyEventZPTFXO: zeptpayWebhook_1.ZeptPayWebhook.verifyEventZPTFXO.bind(zeptpayWebhook_1.ZeptPayWebhook),
            },
        };
        if (this.hasApiKeys) {
            this.logger.log("ZeptPay SDK initialized ✅ (Full API + Webhook mode)");
        }
        else {
            this.logger.log("ZeptPay SDK initialized ✅ (Webhook-only mode - API calls will throw errors)");
        }
    }
    use(plugin) {
        this.plugins.use(plugin, this.client);
        return this;
    }
    get events() {
        return this.client.events;
    }
    get performance() {
        return this.client.performance;
    }
}
exports.ZeptPay = ZeptPay;
exports.default = ZeptPay;
__exportStar(require("./types"), exports);
__exportStar(require("./error/ZeptPayError"), exports);
