// @flixora/zeptpay-payment-core/dist/types/markdown/airxpayTypes.ts
// types/types.ts

// ==================== Config ====================
export interface AirXPayConfig {
  clientKey: string;
  secretKey: string;
  timeout?: number;
  testMode?: boolean;
}

// ==================== Common ====================
export interface PayerInfo {
  userId: string;
  name: string;
  email: string;
}

export interface ErrorResponse {
  success: false;
  error: string;
  code: string;
  timestamp: string;
}

export interface TransactionResponse {
  success: true;
  _id: string;
  airxpayTransactionId: string;
  amount: number;
  currency: string;
  paymentMethod: string;
  status: "captured" | "failed" | "cancelled" | "processing" | "authorized" | "pending";
  paidAt: string | null;
  payer: { name: string };
  receiver: { name: string };
  source: string;
  createdAt: string;
  updatedAt: string;
  message?: string;
  isFinal?: boolean;
}

// ==================== Payments ====================
export interface CreatePaymentRequest {
  vendorCodeUID: string;
  amount: number;
  appName: string;
  currency?: string;
  payer: PayerInfo;
}

export interface ConfirmPaymentRequest {
  txnId: string;
}

export interface GetUserTransactionsParams {
  userId: string;
  role?: "payer" | "merchant" | "all";
  limit?: number;
  skip?: number;
}

export interface GetUserTransactionsResponse {
  success: true;
  count: number;
  transactions: TransactionResponse[];
}

// ==================== QR ====================
export interface GenerateQRRequest {
  vendorCodeUID: string;
  amount: number;
  appName: string;
  currency?: string;
  payer: PayerInfo;
}

export interface GenerateQRResponse extends TransactionResponse {
  qrCodeId: string;
  qrData: string;
  qrPayload: object;
  expiresAt: string;
}

export interface SimulateQrScanRequest {
  qrCodeId: string;
  result: "success" | "failed" | "pending";
}

export interface SimulateQrScanResponse extends TransactionResponse {
  qrCodeId: string;
}

// ==================== AutoPay ====================
export type AutoPayFrequency = "daily" | "weekly" | "monthly";

export interface CreateAutoPayRequest {
  vendorCodeUID: string;
  amount: number;
  appName: string;
  currency?: string;
  payer: PayerInfo;
  frequency?: AutoPayFrequency;
  startDate?: string | Date;
  endDate?: string | Date;
}

export interface CreateAutoPayResponse extends TransactionResponse {
  mandateId: string;
  frequency: AutoPayFrequency;
  nextPaymentDate: string;
}

export interface CaptureAutoPayRequest {
  transactionId: string;
}

export interface CaptureAutoPayResponse extends TransactionResponse {
  nextPaymentDate?: string;
  paymentsProcessed: number;
}
